client/src/main/scala/org/http4s/client/DefaultClient.scala

```diff
@@ -0,0 +1,238 @@
+package org.http4s
+package client
+
+import cats._
+import cats.data.{Kleisli, OptionT}
+import cats.effect._
+import cats.effect.implicits._
+import cats.implicits._
+import fs2._
+import org.http4s.Status.Successful
+import org.http4s.headers.{Accept, MediaRangeAndQValue}
+
+private[client] abstract class DefaultClient[F[_]](implicit F: Bracket[F, Throwable])
+    extends Client[F] {
+  def run(req: Request[F]): Resource[F, Response[F]]
+
+  /** Submits a request, and provides a callback to process the response.
+    *
+    * @param req The request to submit
+    * @param f   A callback for the response to req.  The underlying HTTP connection
+    *            is disposed when the returned task completes.  Attempts to read the
+    *            response body afterward will result in an error.
+    * @return The result of applying f to the response to req
+    */
+  def fetch[A](req: Request[F])(f: Response[F] => F[A]): F[A] =
+    run(req).use(f)
+
+  /** Submits a request, and provides a callback to process the response.
+    *
+    * @param req An effect of the request to submit
+    * @param f A callback for the response to req.  The underlying HTTP connection
+    *          is disposed when the returned task completes.  Attempts to read the
+    *          response body afterward will result in an error.
+    * @return The result of applying f to the response to req
+    */
+  def fetch[A](req: F[Request[F]])(f: Response[F] => F[A]): F[A] =
+    req.flatMap(fetch(_)(f))
+
+  /**
+    * Returns this client as a [[Kleisli]].  All connections created by this
+    * service are disposed on completion of callback task f.
+    *
+    * This method effectively reverses the arguments to `fetch`, and is
+    * preferred when an HTTP client is composed into a larger Kleisli function,
+    * or when a common response callback is used by many call sites.
+    */
+  def toKleisli[A](f: Response[F] => F[A]): Kleisli[F, Request[F], A] =
+    Kleisli(fetch(_)(f))
+
+  @deprecated("Use toKleisli", "0.18")
+  def toService[A](f: Response[F] => F[A]): Service[F, Request[F], A] =
+    toKleisli(f)
+
+  /**
+    * Returns this client as an [[HttpApp]].  It is the responsibility of
+    * callers of this service to run the response body to dispose of the
+    * underlying HTTP connection.
+    *
+    * This is intended for use in proxy servers.  `fetch`, `fetchAs`,
+    * [[toKleisli]], and [[streaming]] are safer alternatives, as their
+    * signatures guarantee disposal of the HTTP connection.
+    */
+  def toHttpApp: HttpApp[F] = Kleisli { req =>
+    def go[A](r: Resource[F, A]): F[(A, F[Unit])] = r match {
+      case Resource.Allocate(r) =>
+        r.map { case (a, release) => (a, release(ExitCase.Completed)) }
+      case Resource.Bind(r, f) =>
```

Comment:
I am sleepy and this feels dangerous.  Please be extra careful reviewing this.

Reply:
😟

Reply:
It probably pays to go back and understand why `Resource` has this ADT instead of being a wrapper around `Allocate`'s value, because we're basically reconstructing that.

(This is also a suggestion for myself.)

Reply:
Specifically, https://github.com/typelevel/cats-effect/pull/278